truffle/src/main/java/org/jruby/truffle/nodes/core/SymbolNodes.java

```diff
@@ -241,10 +240,63 @@ public SwapcaseNode(SwapcaseNode prev) {
         public RubySymbol swapcase(RubySymbol symbol) {
             notDesignedForCompilation();
 
-            ByteList byteList = StringNodes.StringNodesHelper.swapcase(symbol.toRubyString());
+            final ByteList byteList = SymbolNodesHelper.swapcase(symbol.getByteList());
+            return getContext().newSymbol(byteList);
+        }
+
+    }
+
+    @CoreMethod(names = "upcase")
+    public abstract static class UpcaseNode extends CoreMethodNode {
+
+        public UpcaseNode(RubyContext context, SourceSection sourceSection) {
+            super(context, sourceSection);
+        }
+
+        public UpcaseNode(UpcaseNode prev) {
+            super(prev);
+        }
+
+        @Specialization
+        public RubySymbol upcase(RubySymbol symbol) {
+            notDesignedForCompilation();
+
+            final ByteList byteList = SymbolNodesHelper.upcase(symbol.getByteList());
             return getContext().newSymbol(byteList);
         }
 
     }
 
+    public static class SymbolNodesHelper {
+
+        @TruffleBoundary
+        public static ByteList upcase(ByteList originalByteList) {
+            final ByteList byteList = originalByteList.dup();
+            final int length = byteList.length();
+            for (int i = 0; i < length; i++) {
+                final char c = byteList.charAt(i);
+                if ((c >= 'a') && (c <= 'z')) {
+                    byteList.set(i, c & 0x5f);
+                }
+            }
+            return byteList;
+        }
+
+        @TruffleBoundary
+        public static ByteList swapcase(ByteList originalByteList) {
+            final ByteList byteList = originalByteList.dup();
+            final int length = byteList.length();
+            for (int i = 0; i < length; i++) {
+                final char c = byteList.charAt(i);
+                if ((c >= 'a') && (c <= 'z')) {
+                    byteList.set(i, c & 0x5f);
+                } else if ((c >= 'A') && (c <= 'Z')) {
+                    byteList.set(i, c ^ 0x20);
+                }
+            }
+            return byteList;
```

Comment:
Check out this for upcase
https://github.com/jruby/jruby/blob/40784184eaa8a47630aa5de0ccc21f9488ae258e/core/src/main/java/org/jruby/RubySymbol.java#L378
https://github.com/jruby/jruby/blob/40784184eaa8a47630aa5de0ccc21f9488ae258e/core/src/main/java/org/jruby/RubyString.java#L1783